In [109]:
import pandas as pd
import numpy as np
import os
import matplotlib.pyplot as plt

pd.set_option("display.max_columns", None)

In [110]:
DATA_PATH = "../data/raw"

print(os.listdir(DATA_PATH))

['employee_attrition.csv', '.DS_Store', 'essential_skills.csv', 'software_skills.csv', 'hr_performance_engagement.csv', 'occupation_data.csv']


In [111]:
employee_df = pd.read_csv(f"{DATA_PATH}/employee_attrition.csv")
engagement_df = pd.read_csv(f"{DATA_PATH}/hr_performance_engagement.csv")
occupation_df = pd.read_csv(f"{DATA_PATH}/occupation_data.csv")
essential_skills_df = pd.read_csv(f"{DATA_PATH}/essential_skills.csv")
software_skills_df = pd.read_csv(f"{DATA_PATH}/software_skills.csv")

print("Employee:", employee_df.shape)
print("Engagement:", engagement_df.shape)
print("Occupation:", occupation_df.shape)
print("Essential skills:", essential_skills_df.shape)
print("Software skills:", software_skills_df.shape)

Employee: (500, 24)
Engagement: (3150, 39)
Occupation: (1016, 3)
Essential skills: (18200, 15)
Software skills: (31821, 7)


In [112]:
print("EMPLOYEE DATA")
print("=" * 60)
employee_df.head()

EMPLOYEE DATA


,EmployeeID,Name,Gender,Age,Department,JobRole,EducationLevel,JoiningDate,CountryCode,Country,PhoneNumber,MonthlySalary,OvertimeHoursPerMonth,LeavesTaken,LastLeaveDate,LeaveDayName,ProjectsHandled,TrainingHours,CustomerSatisfaction,LastPromotionYear,YearsAtCompany,WorkLifeBalanceScore,PerformanceRating,AttritionRisk
0,1,Steven Barnett,Other,57,Finance,Auditor,2,2016-05-05,91,India,6222371166,93531,33,9,2024-07-03,Wednesday,13,16,NaN,2020,9,1.90,4,No
1,2,Christopher Benson,Female,26,Sales,Sales Executive,2,2014-08-20,1,Canada,466649099,53706,24,1,2024-01-05,Friday,15,44,8.0,2014,11,5.03,1,Yes
2,3,Norman Lane,Other,59,Support,Helpdesk,1,2010-05-17,1,Canada,8302098565,105756,20,7,2024-11-27,Wednesday,7,62,7.0,2012,15,4.83,4,No
3,4,Rita Walker,Female,43,HR,HR Executive,3,2015-06-20,49,Germany,5555799735,176377,22,6,2024-07-15,Monday,15,8,NaN,2016,10,4.60,4,No
4,5,Judith Ware,Male,52,Sales,Account Manager,3,2019-08-20,49,Germany,4796634990,154782,18,16,2024-12-17,Tuesday,10,57,8.0,2022,6,3.73,4,No


In [113]:
employee_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 24 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   EmployeeID             500 non-null    int64  
 1   Name                   500 non-null    str    
 2   Gender                 500 non-null    str    
 3   Age                    500 non-null    int64  
 4   Department             500 non-null    str    
 5   JobRole                500 non-null    str    
 6   EducationLevel         500 non-null    int64  
 7   JoiningDate            500 non-null    str    
 8   CountryCode            500 non-null    int64  
 9   Country                500 non-null    str    
 10  PhoneNumber            500 non-null    int64  
 11  MonthlySalary          500 non-null    int64  
 12  OvertimeHoursPerMonth  500 non-null    int64  
 13  LeavesTaken            500 non-null    int64  
 14  LastLeaveDate          500 non-null    str    
 15  LeaveDayName     

In [114]:
employee_df.isnull().sum().sort_values(ascending=False).head(20)

CustomerSatisfaction     319
EmployeeID                 0
Name                       0
PerformanceRating          0
WorkLifeBalanceScore       0
YearsAtCompany             0
LastPromotionYear          0
TrainingHours              0
ProjectsHandled            0
LeaveDayName               0
LastLeaveDate              0
LeavesTaken                0
OvertimeHoursPerMonth      0
MonthlySalary              0
PhoneNumber                0
Country                    0
CountryCode                0
JoiningDate                0
EducationLevel             0
JobRole                    0
dtype: int64

In [115]:
employee_df.duplicated().sum()

np.int64(0)

In [116]:
employee_df['AttritionRisk'].value_counts(normalize=True) * 100

AttritionRisk
No     89.0
Yes    11.0
Name: proportion, dtype: float64

In [117]:
[c for c in employee_df.columns if 'id' in c.lower()]

['EmployeeID']

In [118]:
print("ENGAGEMENT DATA")
print("=" * 60)
engagement_df.head()

ENGAGEMENT DATA


,Unnamed: 0,FirstName,LastName,StartDate,ExitDate,Title,Supervisor,ADEmail,BusinessUnit,EmployeeStatus,EmployeeType,PayZone,EmployeeClassificationType,TerminationType,TerminationDescription,DepartmentType,Division,DOB,State,JobFunctionDescription,GenderCode,LocationCode,RaceDesc,MaritalDesc,Performance Score,Current Employee Rating,Employee ID,Survey Date,Engagement Score,Satisfaction Score,Work-Life Balance Score,Training Date,Training Program Name,Training Type,Training Outcome,Location,Trainer,Training Duration(Days),Training Cost
0,0,Uriah,Bridges,20-Sep-19,NaN,Production Technician I,Peter Oneill,uriah.bridges@bilearner.com,CCDR,Active,Contract,Zone C,Temporary,Unk,NaN,Production,Finance & Accounting,07-10-1969,MA,Accounting,Female,34904,White,Widowed,Fully Meets,4,3427,14-01-2023,1,2,3,15-Jul-23,Leadership Development,Internal,Failed,South Marisa,Taylor Rodriguez,2,606.11
1,1,Paula,Small,11-Feb-23,NaN,Production Technician I,Renee Mccormick,paula.small@bilearner.com,EW,Active,Contract,Zone A,Part-Time,Unk,NaN,Production,Aerial,30-08-1965,MA,Labor,Male,6593,Hispanic,Widowed,Fully Meets,3,3428,09-09-2022,2,1,5,12-Sep-22,Customer Service,External,Incomplete,Tammieville,Kelly Patterson DDS,4,673.02
2,2,Edward,Buck,10-Dec-18,NaN,Area Sales Manager,Crystal Walker,edward.buck@bilearner.com,PL,Active,Full-Time,Zone B,Part-Time,Unk,NaN,Sales,General - Sga,06-10-1991,MA,Assistant,Male,2330,Hispanic,Widowed,Fully Meets,4,3429,27-05-2023,1,2,1,13-Aug-22,Leadership Development,External,Failed,East Roberthaven,Taylor Thomas,2,413.28
3,3,Michael,Riordan,21-Jun-21,NaN,Area Sales Manager,Rebekah Wright,michael.riordan@bilearner.com,CCDR,Active,Contract,Zone A,Full-Time,Unk,NaN,Sales,Finance & Accounting,04-04-1998,ND,Clerk,Male,58782,Other,Single,Fully Meets,2,3430,16-06-2023,5,5,4,15-Dec-22,Project Management,External,Completed,Garzatown,Holly Elliott,3,663.78
4,4,Jasmine,Onque,29-Jun-19,NaN,Area Sales Manager,Jason Kim,jasmine.onque@bilearner.com,TNS,Active,Contract,Zone A,Temporary,Unk,NaN,Sales,General - Con,29-08-1969,FL,Laborer,Female,33174,Other,Married,Fully Meets,3,3431,25-11-2022,2,5,3,13-Jul-23,Technical Skills,External,Failed,Lake Meganville,Donald Martinez,5,399.03


In [119]:
engagement_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 3150 entries, 0 to 3149
Data columns (total 39 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Unnamed: 0                  3150 non-null   int64  
 1   FirstName                   3150 non-null   str    
 2   LastName                    3150 non-null   str    
 3   StartDate                   3150 non-null   str    
 4   ExitDate                    1606 non-null   str    
 5   Title                       3150 non-null   str    
 6   Supervisor                  3150 non-null   str    
 7   ADEmail                     3150 non-null   str    
 8   BusinessUnit                3150 non-null   str    
 9   EmployeeStatus              3150 non-null   str    
 10  EmployeeType                3150 non-null   str    
 11  PayZone                     3150 non-null   str    
 12  EmployeeClassificationType  3150 non-null   str    
 13  TerminationType             3150 non-null   

In [120]:
engagement_df.isnull().sum().sort_values(ascending=False).head(20)

ExitDate                   1544
TerminationDescription     1544
Unnamed: 0                    0
Satisfaction Score            0
MaritalDesc                   0
Performance Score             0
Current Employee Rating       0
Employee ID                   0
Survey Date                   0
Engagement Score              0
Work-Life Balance Score       0
LocationCode                  0
Training Date                 0
Training Program Name         0
Training Type                 0
Training Outcome              0
Location                      0
Trainer                       0
Training Duration(Days)       0
RaceDesc                      0
dtype: int64

In [121]:
engagement_df.duplicated().sum()

np.int64(150)

In [122]:
[c for c in engagement_df.columns if 'id' in c.lower()]

['Employee ID']

In [123]:
# EmployeeID overlap check between employee_attrition and engagement tables
employee_ids = set(employee_df["EmployeeID"])
engagement_ids = set(engagement_df["Employee ID"])
overlap = employee_ids & engagement_ids
print(f"Employee IDs in employee_attrition: {len(employee_ids)} (range {employee_df['EmployeeID'].min()}-{employee_df['EmployeeID'].max()})")
print(f"Employee IDs in engagement: {len(engagement_ids)} (range {engagement_df['Employee ID'].min()}-{engagement_df['Employee ID'].max()})")
print(f"Overlap: {len(overlap)}")
if len(overlap) == 0:
    print("WARNING: These two tables use different ID spaces and cannot be joined on EmployeeID.")

Employee IDs in employee_attrition: 500 (range 1-500)
Employee IDs in engagement: 3000 (range 1001-4000)
Overlap: 0


In [124]:
print("OCCUPATION DATA")
print("=" * 60)
occupation_df.head()

OCCUPATION DATA


,O*NET-SOC Code,Title,Description
0,11-1011.00,Chief Executives,Determine and formulate policies and provide o...
1,11-1011.03,Chief Sustainability Officers,"Communicate and coordinate with management, sh..."
2,11-1021.00,General and Operations Managers,"Plan, direct, or coordinate the operations of ..."
3,11-1031.00,Legislators,"Develop, introduce, or enact laws and statutes..."
4,11-2011.00,Advertising and Promotions Managers,"Plan, direct, or coordinate advertising polici..."


In [125]:
occupation_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1016 entries, 0 to 1015
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype
---  ------          --------------  -----
 0   O*NET-SOC Code  1016 non-null   str  
 1   Title           1016 non-null   str  
 2   Description     1016 non-null   str  
dtypes: str(3)
memory usage: 23.9 KB


In [126]:
occupation_df.isnull().sum().sort_values(ascending=False).head(20)

O*NET-SOC Code    0
Title             0
Description       0
dtype: int64

In [127]:
occupation_df.duplicated().sum()

np.int64(0)

In [128]:
[c for c in occupation_df.columns if 'id' in c.lower()]

[]

In [129]:
print("ESSENTIAL SKILLS DATA")
print("=" * 60)
essential_skills_df.head()

ESSENTIAL SKILLS DATA


,O*NET-SOC Code,Title,Element ID,Element Name,Scale ID,Scale Name,Data Value,N,Standard Error,Lower CI Bound,Upper CI Bound,Recommend Suppress,Not Relevant,Date,Domain Source
0,11-1011.00,Chief Executives,2.A.1.a,Reading Comprehension,IM,Importance,4.12,8,0.1250,3.8800,4.3700,N,NaN,08/2023,Analyst
1,11-1011.00,Chief Executives,2.A.1.a,Reading Comprehension,LV,Level,4.62,8,0.1830,4.2664,4.9836,N,N,08/2023,Analyst
2,11-1011.00,Chief Executives,2.A.1.b,Active Listening,IM,Importance,4.00,8,0.0000,4.0000,4.0000,N,NaN,08/2023,Analyst
3,11-1011.00,Chief Executives,2.A.1.b,Active Listening,LV,Level,4.75,8,0.1637,4.4292,5.0708,N,N,08/2023,Analyst
4,11-1011.00,Chief Executives,2.A.1.c,Writing,IM,Importance,4.12,8,0.1250,3.8800,4.3700,N,NaN,08/2023,Analyst


In [130]:
essential_skills_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 18200 entries, 0 to 18199
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   O*NET-SOC Code      18200 non-null  str    
 1   Title               18200 non-null  str    
 2   Element ID          18200 non-null  str    
 3   Element Name        18200 non-null  str    
 4   Scale ID            18200 non-null  str    
 5   Scale Name          18200 non-null  str    
 6   Data Value          18200 non-null  float64
 7   N                   18200 non-null  int64  
 8   Standard Error      18200 non-null  float64
 9   Lower CI Bound      18200 non-null  float64
 10  Upper CI Bound      18200 non-null  float64
 11  Recommend Suppress  18200 non-null  str    
 12  Not Relevant        9100 non-null   str    
 13  Date                18200 non-null  str    
 14  Domain Source       18200 non-null  str    
dtypes: float64(4), int64(1), str(10)
memory usage: 2.1 MB


In [131]:
essential_skills_df.isnull().sum().sort_values(ascending=False).head(20)

Not Relevant          9100
O*NET-SOC Code           0
Title                    0
Element ID               0
Element Name             0
Scale ID                 0
Scale Name               0
Data Value               0
N                        0
Standard Error           0
Lower CI Bound           0
Upper CI Bound           0
Recommend Suppress       0
Date                     0
Domain Source            0
dtype: int64

In [132]:
essential_skills_df.duplicated().sum()

np.int64(0)

In [133]:
[c for c in essential_skills_df.columns if 'id' in c.lower()]

['Element ID', 'Scale ID']

In [134]:
print("SOFTWARE SKILLS DATA")
print("=" * 60)
software_skills_df.head()

SOFTWARE SKILLS DATA


,O*NET-SOC Code,Title,Workplace Example,Element ID,Element Name,Hot Technology,In Demand
0,11-1011.00,Chief Executives,Adobe Acrobat,2.E.5.b,Document management software,Y,N
1,11-1011.00,Chief Executives,AdSense Tracker,2.E.6.f,Data base user interface and query software,N,N
2,11-1011.00,Chief Executives,Atlassian JIRA,2.E.5.a,Content workflow software,Y,N
3,11-1011.00,Chief Executives,Blackbaud The Raiser's Edge,2.E.6.c,Customer relationship management CRM software,N,N
4,11-1011.00,Chief Executives,ComputerEase construction accounting software,2.E.2.a,Accounting software,N,N


In [135]:
software_skills_df.info()

<class 'pandas.DataFrame'>
RangeIndex: 31821 entries, 0 to 31820
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   O*NET-SOC Code     31821 non-null  str  
 1   Title              31821 non-null  str  
 2   Workplace Example  31821 non-null  str  
 3   Element ID         31821 non-null  str  
 4   Element Name       31821 non-null  str  
 5   Hot Technology     31821 non-null  str  
 6   In Demand          31821 non-null  str  
dtypes: str(7)
memory usage: 1.7 MB


In [136]:
software_skills_df.isnull().sum().sort_values(ascending=False).head(20)

O*NET-SOC Code       0
Title                0
Workplace Example    0
Element ID           0
Element Name         0
Hot Technology       0
In Demand            0
dtype: int64

In [137]:
software_skills_df.duplicated().sum()

np.int64(0)

In [138]:
[c for c in software_skills_df.columns if 'id' in c.lower()]

['Element ID']

# End-of-Day Data Understanding Report

## 1. employee_attrition.csv

- **Row/column count:** 500 rows × 24 columns
- **Likely primary key:** `EmployeeID`
- **Missing values:** `CustomerSatisfaction` has 319 missing values. All other columns have 0 missing values.
- **Purpose:** Feeds the attrition ML model directly.
- **Potential join:** `EmployeeID` may connect to `hr_performance_engagement.csv` through `Employee ID`.

## 2. hr_performance_engagement.csv

- **Row/column count:** 3,150 rows × 39 columns
- **Likely primary key:** `Employee ID`
- **Missing values:** `ExitDate` has 1,544 missing values and `TerminationDescription` has 1,544 missing values. All other columns have 0 missing values.
- **Purpose:** Engagement analytics; no ML needed initially.
- **Potential join:** `Employee ID` may connect to `employee_attrition.csv` through `EmployeeID`.

## 3. occupation_data.csv

- **Row/column count:** 1,016 rows × 3 columns
- **Likely primary key:** No obvious ID column was identified.
- **Missing values:** No missing values.
- **Purpose:** Becomes the role master reference table.
- **Potential join:** `O*NET-SOC Code` may connect this dataset to the skills datasets.

## 4. essential_skills.csv

- **Row/column count:** 18,200 rows × 15 columns
- **Likely primary key:** No single primary key was established from the ID-column check.
- **Missing values:** No missing values.
- **Purpose:** Provides essential-skills information for the required-skills data per role.
- **Potential join:** `O*NET-SOC Code` may connect this dataset to `occupation_data.csv`.

## 5. software_skills.csv

- **Row/column count:** 31,821 rows × 7 columns
- **Likely primary key:** No single primary key was established from the ID-column check.
- **Missing values:** No missing values.
- **Purpose:** Provides software-skills information for the required-skills data per role.
- **Potential join:** `O*NET-SOC Code` may connect this dataset to `occupation_data.csv`.

## Overall Data Understanding Conclusion

All five datasets have been loaded and inspected for shape, columns, data types, missing values, duplicate rows, and candidate ID columns.

The main candidate employee relationship is:

`employee_attrition.csv.EmployeeID` ↔ `hr_performance_engagement.csv.Employee ID`

The occupation and skills datasets contain `O*NET-SOC Code` as a potential relationship.